lib/internal/modules/cjs/loader.js

```diff
@@ -746,6 +746,15 @@ Module.runMain = function() {
   process._tickCallback();
 };
 
+Module.createRequireFunction = (filename) => {
```

Comment:
Can this be made to only take URLs for now so that we don't have to deal w/ path vs URL string collisions? And/or made to accept URL strings and fallback to path?

Reply:
or only paths? node doesn't use urls.

Reply:
`fs` supports urls

Reply:
but `require` does not, and this is analogous to requiring modules, not reading from the filesystem.

Reply:
I'm 👎 for only `URL`s here.

Reply:
@zenparsing would it be fine for it to be just file paths?

Reply:
Does `new Module("file://...")` work? If not then I suppose that there's no easy way to use this with `import.meta.url`.

Reply:
@zenparsing 
`new Module('literally anything')` works, but that's not the point. If we only accept a string here, it has to either be a URL or a file path. It can't be both, and I want to be able to use file paths here.

Reply:
so ... back to the _not community friendly_ again ... can we have this **and** `import.meta.require` that works out of the box too, pretty please?

Reply:
could it be a file path string, or a `URL` object? That would allow for both, with a clear separation between the two (`import.meta.url` could be a `URL` object, for example)

Reply:
@WebReflection this api isn't stopping from `import.meta.require` from existing or not existing.

Reply:
@ljharb that is the current API. the issue is that @zenparsing thinks that `createRequireFunction(new URL(import.meta.url))` is to much.

Reply:
Why not just add extra static constructors a la `Buffer` - `Module.fromPath("/foo/bar")`, `Module.fromURL(import.meta.url)`, and so on?

Reply:
i'd be okay with `createRequireFromURL` and `createRequireFromFilePath`

Reply:
could `import.meta.url` be an immutable/frozen URL object, or is that not possible?

Reply:
Note it's more important to note the "file" aspect in the URL case than the path case - `createRequireFromFileURL` and `createRequireFromPath`.

Reply:
> createRequireFromFileURL

at this point call it `createRequireFromMetaURL` ?

Reply:
it's not just for import.meta.url

Reply:
Why is that needed ? Could it not be handled both `createRequire(url || path)` without adding two extra API's for just that ? The naming is very verbose...

Reply:
@michael-ciniawsky is `./foo` or `/foo` or `../foo` a URL, or a relative file path?

Reply:
`/foo` is an absolute path
`./foo && ../foo` are relative paths


?

Reply:
they're also both valid URLs, since they would work in `<a href="">` - unless we're saying that a URL requires a protocol?

Reply:
Yeah, a URL should include the protocol how should it be (reliably) possible to know it's a URL otherwise?

Reply:
`createRequire(..., { url: true})` instead , if non protocol URL's _really_ need to be supported? But I don't get it... :)

Reply:
they are valid URLs though. we don't get to put arbitrary decisions on specified behaviour

Reply:
Does this depend on the context (in the browser) or in general ? And if e.g `/foo` is a valid URL and also a valid file path what does this mean for `node` in particular? It's treaten as a path ?

Reply:
@michael-ciniawsky URLs have associated string encodings and formats such as [Relative URL Strings](https://url.spec.whatwg.org/#relative-url-string).

In particular URLs and file paths have different encodings and parsing so there are important differences in *what* is resolved when interpretting them:

string | file from file path | file from URL pathname (as a Relative URL String, with cannot-be-base-url flag) |
--- | --- | ---
* `C:\foo.txt` | `C:\foo.txt` | invalid
`./foo%20bar.txt` | `./foo%20bar.txt` | `./foo bar.txt`
* `/foo#bar` | `/foo#bar` | `/foo`

We can see differences in validity, differences in encoding, and differences in parsing all in the above examples. This is why we need to be specific about which format we are talking about.

Reply:
@zenparsing how does that work for `data:` or `blob:` URLs?